File: sdk/typescript/src/instrumentor.ts
/**
 * HikariInstrumentor: auto-instruments available LLM provider clients.
 *
 * Tries to patch OpenAI, Anthropic, and Google clients. Skips providers
 * that are not installed or have incompatible versions.
 */

import type { PricingModel } from "./pricing.js";
import * as openai from "./providers/openai.js";
import * as anthropic from "./providers/anthropic.js";
import * as google from "./providers/google.js";

/**
 * Manages auto-instrumentation of LLM provider clients.
 */
export class HikariInstrumentor {
  private patchedProviders: string[] = [];

  /**
   * Apply monkey-patches to all available provider clients.
   *
   * @param pricing - The pricing model to use for cost computation
   * @returns List of successfully patched provider names
   */
  instrument(pricing: PricingModel): string[] {
    this.patchedProviders = [];

    if (openai.patch(pricing)) {
      this.patchedProviders.push("openai");
    }

    if (anthropic.patch(pricing)) {
      this.patchedProviders.push("anthropic");
    }

    if (google.patch(pricing)) {
      this.patchedProviders.push("google");
    }

    return [...this.patchedProviders];
  }

  /**
   * Restore all patched provider clients to their original state.
   */
  uninstrument(): void {
    if (this.patchedProviders.includes("openai")) {
      openai.unpatch();
    }

    if (this.patchedProviders.includes("anthropic")) {
      anthropic.unpatch();
    }

    if (this.patchedProviders.includes("google")) {
      google.unpatch();
    }

    this.patchedProviders = [];
  }

  /**
   * Get the list of currently patched providers.
   *
   * @returns Array of patched provider names
   */
  getPatchedProviders(): string[] {
    return [...this.patchedProviders];
  }
}

File: sdk/python/src/hikari/instrumentor.py
"""Auto-patching orchestrator.

Detects installed provider libraries and applies monkey-patches to capture
LLM call telemetry as OTel spans with ``hikari.*`` attributes.
"""

from __future__ import annotations

import logging
from typing import Any

from hikari.pricing import PricingModel
from hikari.providers import anthropic as anthropic_patch
from hikari.providers import google as google_patch
from hikari.providers import openai as openai_patch

logger = logging.getLogger("hikari.instrumentor")

_PATCHES = [
    ("openai", openai_patch),
    ("anthropic", anthropic_patch),
    ("google", google_patch),
]


class HikariInstrumentor:
    """Orchestrates monkey-patching of provider clients."""

    def __init__(self, pricing_model: PricingModel) -> None:
        self._pricing_model = pricing_model
        self._patched: list[str] = []

    def instrument(self) -> None:
        """Detect installed providers and patch them.

        For each provider:
        1. Try to import the provider module.
        2. If import fails -> skip (provider not installed).
        3. If import succeeds, check version meets minimum.
        4. If version too old -> log warning, skip.
        5. If version OK -> apply patch.
        """
        for name, patch_module in _PATCHES:
            try:
                success = patch_module.patch(self._pricing_model)
                if success:
                    self._patched.append(name)
            except Exception:
                logger.warning("Failed to instrument %s", name, exc_info=True)

    def uninstrument(self) -> None:
        """Restore all original methods."""
        for name, patch_module in _PATCHES:
            if name in self._patched:
                try:
                    patch_module.unpatch()
                except Exception:
                    logger.warning("Failed to uninstrument %s", name, exc_info=True)
        self._patched.clear()

    @property
    def patched_providers(self) -> list[str]:
        """Return list of successfully patched provider names."""
        return list(self._patched)
